// openrouteservice-js/npm_distribution/OrsElevation.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _superagent = require('superagent');

var _superagent2 = _interopRequireDefault(_superagent);

var _bluebird = require('bluebird');

var _bluebird2 = _interopRequireDefault(_bluebird);

var _OrsUtil = require('./OrsUtil');

var _OrsUtil2 = _interopRequireDefault(_OrsUtil);

var _constants = require('./constants');

var _constants2 = _interopRequireDefault(_constants);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var orsUtil = new _OrsUtil2.default();

var OrsElevation = function () {
  function OrsElevation(args) {
    _classCallCheck(this, OrsElevation);

    this.args = {};
    if (_constants2.default.propNames.apiKey in args) {
      this.args[_constants2.default.propNames.apiKey] = args[_constants2.default.propNames.apiKey];
    } else {
      // eslint-disable-next-line no-console
      console.error(_constants2.default.missingAPIKeyMsg);
    }
  }

  _createClass(OrsElevation, [{
    key: 'clear',
    value: function clear() {
      for (var variable in this.args) {
        if (variable !== _constants2.default.propNames.apiKey) delete this.args[variable];
      }
    }
  }, {
    key: 'generatePayload',
    value: function generatePayload(args) {
      var payload = {};

      for (var key in args) {
        if (_constants2.default.baseUrlConstituents.indexOf(key) > -1) {
          continue;
        } else {
          payload[key] = args[key];
        }
      }
      return payload;
    }
  }, {
    key: 'elevationPromise',
    value: function elevationPromise() {
      var that = this;

      return new _bluebird2.default(function (resolve, reject) {
        var timeout = 5000;

        var url = orsUtil.prepareUrl(that.args);

        var payload = that.generatePayload(that.args);
        var authorization = that.args[_constants2.default.propNames.apiKey];
        var orsRequest = _superagent2.default.post(url).send(payload).accept(that.args[_constants2.default.propNames.mimeType]).set('Authorization', authorization).timeout(timeout);
        // .accept(that.meta.mimeType)

        for (var key in that.customHeaders) {
          orsRequest.set(key, that.customHeaders[key]);
        }
        orsRequest.end(function (err, res) {
          if (err || !res.ok) {
            // eslint-disable-next-line no-console
            console.error(err);
            reject(err);
          } else if (res) {
            resolve(res.body || res.text);
          }
        });
      });
    }
  }, {
    key: 'lineElevation',
    value: function lineElevation(reqArgs) {
      this.customHeaders = [];
      if (reqArgs.customHeaders) {
        this.customHeaders = reqArgs.customHeaders;
        delete reqArgs.customHeaders;
      }
      orsUtil.setRequestDefaults(this.args, reqArgs);
      // eslint-disable-next-line prettier/prettier
      if (!this.args[_constants2.default.propNames.service] && !reqArgs[_constants2.default.propNames.service]) {
        reqArgs[[_constants2.default.propNames.service]] = 'elevation/line';
      }
      orsUtil.copyProperties(reqArgs, this.args);
      return this.elevationPromise();
    }
  }, {
    key: 'pointElevation',
    value: function pointElevation(reqArgs) {
      // Get custom header and remove from args
      this.customHeaders = [];
      if (reqArgs.customHeaders) {
        this.customHeaders = reqArgs.customHeaders;
        delete reqArgs.customHeaders;
      }

      orsUtil.setRequestDefaults(this.args, reqArgs);
      // eslint-disable-next-line prettier/prettier
      if (!this.args[_constants2.default.propNames.service] && !reqArgs[_constants2.default.propNames.service]) {
        reqArgs[[_constants2.default.propNames.service]] = 'elevation/point';
      }

      orsUtil.copyProperties(reqArgs, this.args);
      return this.elevationPromise();
    }
  }]);

  return OrsElevation;
}();

exports.default = OrsElevation;
