// openrouteservice-js/npm_distribution/OrsPois.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _superagent = require('superagent');

var _superagent2 = _interopRequireDefault(_superagent);

var _bluebird = require('bluebird');

var _bluebird2 = _interopRequireDefault(_bluebird);

var _OrsUtil = require('./OrsUtil');

var _OrsUtil2 = _interopRequireDefault(_OrsUtil);

var _constants = require('./constants');

var _constants2 = _interopRequireDefault(_constants);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var orsUtil = new _OrsUtil2.default();

var OrsPois = function () {
  function OrsPois(args) {
    _classCallCheck(this, OrsPois);

    this.args = {};
    if (_constants2.default.propNames.apiKey in args) {
      this.args[_constants2.default.propNames.apiKey] = args[_constants2.default.propNames.apiKey];
    } else {
      // eslint-disable-next-line no-console
      console.error(_constants2.default.missingAPIKeyMsg);
    }
    if (_constants2.default.propNames.host in args) {
      this.args[_constants2.default.propNames.host] = args[_constants2.default.propNames.host];
    }
    if (_constants2.default.propNames.service in args) {
      this.args[_constants2.default.propNames.service] = args[_constants2.default.propNames.service];
    }
  }

  _createClass(OrsPois, [{
    key: 'clear',
    value: function clear() {
      for (var variable in this.args) {
        if (variable !== _constants2.default.propNames.apiKey) delete this.args[variable];
      }
    }
  }, {
    key: 'generatePayload',
    value: function generatePayload(args) {
      var payload = {};

      for (var key in args) {
        if (_constants2.default.baseUrlConstituents.indexOf(key) > -1) {
          continue;
        } else {
          payload[key] = args[key];
        }
      }
      return payload;
    }
  }, {
    key: 'poisPromise',
    value: function poisPromise() {
      // the service arg is used to build the target url
      if (!this.args[_constants2.default.propNames.service]) {
        this.args[_constants2.default.propNames.service] = 'pois';
      }
      // the request arg is required by the API as part of the body
      this.args.request = this.args.request || 'pois';

      var that = this;
      return new _bluebird2.default(function (resolve, reject) {
        var timeout = 5000;

        var url = orsUtil.prepareUrl(that.args);

        url += url.indexOf('?') > -1 ? '&' : '?';

        if (that.args[_constants2.default.propNames.service]) {
          delete that.args[_constants2.default.propNames.service];
        }

        var payload = that.generatePayload(that.args);
        var authorization = that.args[_constants2.default.propNames.apiKey];

        var orsRequest = _superagent2.default.post(url).send(payload).set('Authorization', authorization).timeout(timeout);

        for (var key in that.customHeaders) {
          orsRequest.set(key, that.customHeaders[key]);
        }
        orsRequest.end(function (err, res) {
          if (err || !res.ok) {
            // eslint-disable-next-line no-console
            console.error(err);
            reject(err);
          } else if (res) {
            resolve(res.body || res.text);
          }
        });
      });
    }
  }, {
    key: 'pois',
    value: function pois(reqArgs) {
      // Get custom header and remove from args
      this.customHeaders = [];
      if (reqArgs.customHeaders) {
        this.customHeaders = reqArgs.customHeaders;
        delete reqArgs.customHeaders;
      }
      orsUtil.setRequestDefaults(this.args, reqArgs);
      orsUtil.copyProperties(reqArgs, this.args);

      return this.poisPromise();
    }
  }]);

  return OrsPois;
}();

exports.default = OrsPois;
